lib/Onyx.d.ts

```diff
@@ -0,0 +1,315 @@
+};
+
+/**
+ * Represents the base options used in `Onyx.connect()` method.
+ */
+type BaseConnectOptions = {
+    statePropertyName?: string;
+    withOnyxInstance?: Component;
+    initWithStoredValues?: boolean;
+};
+
+/**
+ * Represents the options used in `Onyx.connect()` method.
+ * The type is built from `BaseConnectOptions` and extended to handle key/callback related options.
+ * It includes two different forms, depending on whether we are waiting for a collection callback or not.
+ *
+ * If `waitForCollectionCallback` is `true`, it expects `key` to be a Onyx collection key and `callback` will pass `value` as an `OnyxCollection`.
+ *
+ * If `waitForCollectionCallback` is `false` or not specified, the `key` can be any Onyx key and `callback` will pass `value` as an `OnyxEntry`.
+ */
+type ConnectOptions<TKey extends OnyxKey> = BaseConnectOptions &
+    (
+        | {
+              key: TKey extends CollectionKey ? TKey : never;
+              callback?: (value: OnyxCollection<KeyValueMapping[TKey]>, key?: TKey) => void;
+              waitForCollectionCallback: true;
+          }
+        | {
+              key: TKey;
+              callback?: (value: OnyxEntry<KeyValueMapping[TKey]>, key?: TKey) => void;
+              waitForCollectionCallback?: false;
+          }
+    );
+
+/**
+ * Represents a mapping between Onyx collection keys and their respective values.
+ * It helps to enforce that a Onyx collection key should not be without suffix (e.g. should always be of the form `${TKey}${string}`),
+ * and to map each Onyx collection key with suffix to a value of type `TValue`.
+ */
+type Collection<TKey extends CollectionKeyBase, TMap, TValue> = {
+    [MapK in keyof TMap]: MapK extends `${TKey}${string}`
+        ? MapK extends `${TKey}`
+            ? never // forbids empty id
+            : TValue
+        : never;
+};
+
+/**
+ * Represents different kinds of updates that can be passed to `Onyx.update()` method. It is a discriminated union of
+ * different update methods (`SET`, `MERGE`, `MERGE_COLLECTION`), each with their own key and value structure.
+ */
+type OnyxUpdate =
+    | {
+          [TKey in OnyxKey]:
+              | {
+                    onyxMethod: typeof METHOD.SET;
+                    key: TKey;
+                    value: OnyxEntry<KeyValueMapping[TKey]>;
+                }
+              | {
+                    onyxMethod: typeof METHOD.MERGE;
+                    key: TKey;
+                    value: PartialDeep<KeyValueMapping[TKey]>;
+                };
+      }[OnyxKey]
+    | {
+          [TKey in CollectionKeyBase]: {
+              onyxMethod: typeof METHOD.MERGE_COLLECTION;
+              key: TKey;
+              value: Record<`${TKey}${string}`, PartialDeep<KeyValueMapping[TKey]>>;
+          };
+      }[CollectionKeyBase];
+
+/**
+ * Represents the options used in `Onyx.init()` method.
+ */
+type InitOptions = {
+    keys?: DeepRecord<string, string>;
```

Comment:
Is `key` optional because `init` can be called in multiple places? Sorry I know this question is just to increase my Onyx knowledge 😓

Reply:
@hayata-suenaga Adjusted!